tree/tree/src/TBasket.cxx

```diff
@@ -725,20 +740,48 @@ void TBasket::Reset()
          }
       }
    }
-   /*
-      Philippe has asked us to keep this turned off until we finish memory fragmentation studies.
-   // If fBufferRef grew since we last saw it, shrink it to 105% of the occupied size
-   if (curSize > fLastWriteBufferSize) {
-      if (newSize == -1) {
-         newSize = Int_t(1.05*Float_t(fBufferRef->Length()));
+   // If fBufferRef grew since we last saw it, shrink it to "target memory ratio" of the occupied size
+   // This discourages us from having poorly-occupied buffers on branches with little variability.
+   //
+   // Does not help protect against a burst in event sizes, but does help in the cases where the basket
+   // size jumps from 4MB to 8MB while filling the basket, but we only end up utilizing 4.1MB.
+   //
+   // The above code block is meant to protect against extremely large events.
+
+   Float_t target_mem_ratio = fBranch->GetTree()->GetTargetMemoryRatio();
+   Int_t max_size = std::max(fLastWriteBufferSize[0], std::max(fLastWriteBufferSize[1], fLastWriteBufferSize[2]));
+   Int_t target_size = static_cast<Int_t>(target_mem_ratio * Float_t(max_size));
+   if (max_size && (curSize > target_size) && (newSize == -1)) {
+      newSize = target_size;
+      newSize = newSize + 512 - newSize % 512; // Wiggle room and alignment, as above.
+      // We only bother with a resize if it saves 8KB (two normal memory pages).
+      if ((newSize > curSize - 8 * 1024) ||
+          (static_cast<Float_t>(curSize) / static_cast<Float_t>(newSize) < target_mem_ratio)) {
+         newSize = -1;
+      } else if (gDebug > 0) {
+         Info("TBasket::Reset", "Resizing to %ld bytes (was %d); last three sizes were [%d, %d, %d].", newSize, curSize,
```

Comment:
use 'just' Info("Reset"), since Info is a member function of TObject it already knows the class name.